main.cpp

```diff
@@ -13,145 +11,295 @@
+{}
+
+
+
+std::vector<std::vector<double>> Sample::return_data() const
 {
-  // data used for likelihood
-  const Data::PomData raw_data;
-  const std::vector<std::valarray<double>> data_atoms = { std::pow(raw_data.tem_diam_time1/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time2/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time3/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time4/0.3000805, 3)};
+  return const_parameters.data_size;
+}
 
-  // create model
-  Models::TwoStepAlternative model;
 
-  // set up initial conditions and times for the ODE solution
-  std::valarray<double> initialCondition(0.0,prm.n_variables);
-  initialCondition[0] = 0.0012;
-  const std::vector<double> times = {0,
-                                     raw_data.tem_time1,
-                                     raw_data.tem_time2,
-                                     raw_data.tem_time3,
-                                     raw_data.tem_time4};
 
+std::vector<double> Sample::return_times() const
+{
+  return const_parameters.times;
+}
 
-  // define parameters for the histogram
-  const Histograms::Parameters histogramParameters(25, model.getSmallestParticleSize(prm), model.getLargestParticleSize(prm));
 
-  // calculate log likelihood and return it
-  return Statistics::log_likelihood(data_atoms, times, model, prm, initialCondition, histogramParameters);
 
+// Here we implement a Four-Step Model
+Model::Model Sample::return_model() const
+{
+  Model::TermolecularNucleation nucleation(const_parameters.A_index, const_parameters.As_index,
+                                           const_parameters.ligand_index, const_parameters.min_size,
+                                           const_parameters.kf, kb, k1, const_parameters.solvent);
+  Model::Growth small_growth(const_parameters.A_index, const_parameters.min_size, cutoff,
+                             const_parameters.max_size, const_parameters.ligand_index,
+                             const_parameters.conserved_size, k2);
+  Model::Growth large_growth(const_parameters.A_index, cutoff+1, const_parameters.max_size,
+                             const_parameters.max_size, const_parameters.ligand_index,
+                             const_parameters.conserved_size, k3);
+  Model::Agglomeration agglomeration(const_parameters.min_size, cutoff,
+                                     const_parameters.min_size, cutoff,
+                                     const_parameters.max_size, const_parameters.conserved_size,
+                                     k4);
+
+  Model::Model model(const_parameters.min_size, const_parameters.max_size);
+  model.add_rhs_contribution(nucleation);
+  model.add_rhs_contribution(small_growth);
+  model.add_rhs_contribution(large_growth);
+  model.add_rhs_contribution(agglomeration);
+
+  return model;
```

Comment:
The 'model' variable dies here, but before it dies it is copied into the location where you're doing the assignment to in the call to `return_model()`. So the pointer survives, but the object pointed to doesn't. This is not going to work. You need to make sure that the rhs object pointed to survives as long as necessary (namely, to the end of all copies of the model variable) and is destroyed when the last model object that points to it dies itself. The class you're looking for is called `std::shared_ptr`. I'll write you a patch in a minute.